stackrox/stackrox
# Write frontend credential expiry proto and rewrite services in TypeScript

## Description

### Goals

* Increase type safety with compile-time checks.
* Simplify maintenance of consistent backend and frontend response types.

### Changed files

1. Edit cypress/integration/certexpiration.test.js
    * Replace `cy.visit` with `visitMainDashboard` helper function

2. Rename and edit src/Containers/CredentialExpiryBanners/CredentialExpiry.js
    * Add `nameOfComponent` to compute name because `component` prop value becomes non-presentational
    * Replace `expiryFetchFunc` prop with `fetchCertExpiryForComponent` function
    * Replace `userRolePermissions` with `hasServiceIdentityWritePermission` prop
    * Replace `downloadYAMLFunc` prop with `generateCertSecretForComponent` function

3. Delete src/Containers/CredentialExpiryBanners/CredentialExpiryBanners.js

4. Edit src/Containers/MainPage/MainPage.tsx
    * Replace `CredentialExpiryBanners` element with `CredentialExpiryBanner` elements

5. Rename and edit src/services/CertGenerationService.js
    * Import `CertExpiryComponent` type
    * Replace specific `generateCentralCertSecret` and `generateScannerCertSecret` with generic `generateCertSecretForComponent` function

6. Rename and edit src/services/CredentialExpiryService.js
    * Import `CertExpiryComponent` type
    * Replace specific `fetchCentralCertExpiry` and `fetchScannerCertExpiry` with generic `fetchCertExpiryForComponent` function

7. Add src/types/credentialExpiryService.proto.ts which corresponds to https://github.com/stackrox/stackrox/blob/master/proto/api/v1/credential_expiry_service.proto

## Checklist

## Testing Performed

### manual testing

1. `yarn build` in ui
2. `yarn start` in ui

    * See same requests and returned values in banner components (with temporary console.log)

    * See same request to download secret file for CENTRAL and file in bottom bar (with temporary change to expiry)
        ![certgen-CENTRAL-2](https://user-images.githubusercontent.com/11862657/169371179-c86203a1-095f-4ca1-adf6-d37a47b1143c.png)

    * See same request to download secret file for SCANNER and file in bottom bar (with temporary change to expiry)
        ![certgen-SCANNER-2](https://user-images.githubusercontent.com/11862657/169371409-55a8f6bc-c25e-4b75-8962-96ad7f3b59e2.png)

### integration testing

1. `yarn cypress-open` in ui/apps/platform
    * certexpiration.test.js
